File: src/omarcs/spray.py
from __future__ import annotations

import math
import statistics
from collections import defaultdict
from typing import Any

import polars as pl

from .geometry import visible_rows

SPRAY_GAP_TICKS = 16
MIN_SPRAY_SHOTS = 5
MAX_SPRAY_SHOTS = 10
MAX_TARGET_ANGLE = 15.0
MAX_PROJECTED_OFFSET = 220.0

WEAPONS = {
    7: {"id": "ak47", "name": "AK-47", "shortName": "AK", "maxSpeed": 215},
    13: {"id": "galilar", "name": "Galil AR", "shortName": "GALIL", "maxSpeed": 215},
    16: {"id": "m4a4", "name": "M4A4", "shortName": "M4A4", "maxSpeed": 225},
    60: {
        "id": "m4a1_silencer",
        "name": "M4A1-S",
        "shortName": "M4A1-S",
        "maxSpeed": 225,
    },
}
WEAPON_ORDER = tuple(weapon["id"] for weapon in WEAPONS.values())
WEAPONS_BY_ID = {weapon["id"]: weapon for weapon in WEAPONS.values()}


def rows(frame: Any) -> list[dict[str, Any]]:
    if frame is None:
        return []
    if hasattr(frame, "to_dicts"):
        return frame.to_dicts()
    return list(frame)


def player_id(raw: Any) -> str:
    if raw is None:
        return ""
    if isinstance(raw, float) and raw.is_integer():
        return str(int(raw))
    return str(raw)


def vector_dot(
    first: tuple[float, float, float], second: tuple[float, float, float]
) -> float:
    return sum(left * right for left, right in zip(first, second))


def vector_subtract(
    first: tuple[float, float, float], second: tuple[float, float, float]
) -> tuple[float, float, float]:
    return tuple(left - right for left, right in zip(first, second))  # type: ignore[return-value]


def vector_add_scaled(
    origin: tuple[float, float, float],
    direction: tuple[float, float, float],
    scale: float,
) -> tuple[float, float, float]:
    return tuple(
        start + component * scale for start, component in zip(origin, direction)
    )  # type: ignore[return-value]


def vector_length(vector: tuple[float, float, float]) -> float:
    return math.sqrt(vector_dot(vector, vector))


def normalize(vector: tuple[float, float, float]) -> tuple[float, float, float] | None:
    length = vector_length(vector)
    if length <= 1e-9:
        return None
    return tuple(component / length for component in vector)  # type: ignore[return-value]


def cross(
    first: tuple[float, float, float], second: tuple[float, float, float]
) -> tuple[float, float, float]:
    return (
        first[1] * second[2] - first[2] * second[1],
        first[2] * second[0] - first[0] * second[2],
        first[0] * second[1] - first[1] * second[0],
    )


def direction_from_angles(
    pitch_degrees: float, yaw_degrees: float
) -> tuple[float, float, float]:
    pitch = math.radians(pitch_degrees)
    yaw = math.radians(yaw_degrees)
    return (
        math.cos(pitch) * math.cos(yaw),
        math.cos(pitch) * math.sin(yaw),
        -math.sin(pitch),
    )


def angular_distance(
    first: tuple[float, float, float], second: tuple[float, float, float]
) -> float:
    return math.degrees(math.acos(max(-1.0, min(1.0, vector_dot(first, second)))))


def shot_origin(shot: dict[str, Any]) -> tuple[float, float, float] | None:
    try:
        origin = (
            float(shot["origin_x"]),
            float(shot["origin_y"]),
            float(shot["origin_z"]),
        )
    except (KeyError, TypeError, ValueError):
        return None
    return origin if all(math.isfinite(value) for value in origin) else None


def shot_direction(shot: dict[str, Any]) -> tuple[float, float, float] | None:
    try:
        pitch = float(shot["angles_x"])
        yaw = float(shot["angles_y"])
    except (KeyError, TypeError, ValueError):
        return None
    if not math.isfinite(pitch) or not math.isfinite(yaw):
        return None
    return direction_from_angles(pitch, yaw)


def target_head(target: dict[str, Any]) -> tuple[float, float, float]:
    duck = float(target.get("target_duck") or 0.0)
    return (
        float(target["target_x"]),
        float(target["target_y"]),
        float(target["target_z"]) + 64.0 - 18.0 * duck,
    )


def projected_offset(
    shot: dict[str, Any], target: dict[str, Any]
) -> tuple[float, float] | None:
    origin = shot_origin(shot)
    direction = shot_direction(shot)
    if origin is None or direction is None:
        return None
    head = target_head(target)
    target_vector = vector_subtract(head, origin)
    normal = normalize(target_vector)
    if normal is None:
        return None
    denominator = vector_dot(direction, normal)
    if denominator <= 0.05:
        return None
    impact = vector_add_scaled(
        origin, direction, vector_length(target_vector) / denominator
    )
    offset = vector_subtract(impact, head)
    right = normalize((-normal[1], normal[0], 0.0))
    if right is None:
        return None
    up = normalize(cross(normal, right))
    if up is None:
        return None
    horizontal = vector_dot(offset, right)
    vertical = vector_dot(offset, up)
    if not math.isfinite(horizontal) or not math.isfinite(vertical):
        return None
    if abs(horizontal) > MAX_PROJECTED_OFFSET or abs(vertical) > MAX_PROJECTED_OFFSET:
        return None
    return horizontal, vertical


def fire_bullet_rows(demo: Any, steam_id: str) -> list[dict[str, Any]]:
    events = getattr(demo, "events", {}) or {}
    bullets = rows(events.get("fire_bullets"))
    return sorted(
        (row for row in bullets if player_id(row.get("user_steamid")) == steam_id),
        key=lambda row: int(row.get("tick") or 0),
    )


def group_sprays(
    bullets: list[dict[str, Any]],
) -> list[tuple[dict[str, Any], list[dict[str, Any]]]]:
    grouped: list[tuple[dict[str, Any], list[dict[str, Any]]]] = []
    current: list[dict[str, Any]] = []
    current_weapon: dict[str, Any] | None = None
    previous_tick: int | None = None

    def finish() -> None:
        nonlocal current, current_weapon
        if current_weapon is not None and len(current) >= MIN_SPRAY_SHOTS:
            grouped.append((current_weapon, current[:MAX_SPRAY_SHOTS]))
        current = []
        current_weapon = None

    for bullet in bullets:
        try:
            item_definition = int(bullet.get("item_def_index"))
            tick = int(bullet.get("tick"))
        except (TypeError, ValueError):
            finish()
            previous_tick = None
            continue
        weapon = WEAPONS.get(item_definition)
        if weapon is None:
            finish()
            previous_tick = None
            continue
        continues = (
            current_weapon is not None
            and current_weapon["id"] == weapon["id"]
            and previous_tick is not None
            and tick - previous_tick <= SPRAY_GAP_TICKS
        )
        if not continues:
            finish()
            current_weapon = weapon
        current.append(bullet)
        previous_tick = tick
    finish()
    return grouped


def target_candidates(
    ticks: pl.DataFrame,
    steam_id: str,
    shot_ticks: set[int],
    mesh: Any,
) -> dict[int, list[dict[str, Any]]]:
    required = {
        "tick",
        "steamid",
        "side",
        "health",
        "pitch",
        "yaw",
        "X",
        "Y",
        "Z",
        "duck_amount",
        "velocity",
        "approximate_spotted_by",
    }
    if not required.issubset(ticks.columns) or not shot_ticks:
        return {}
    selected = ticks.filter(pl.col("tick").is_in(shot_ticks)).to_dicts()
    viewers: dict[int, dict[str, Any]] = {}
    enemies: dict[int, list[dict[str, Any]]] = defaultdict(list)
    for row in selected:
        tick = int(row.get("tick") or 0)
        if player_id(row.get("steamid")) == steam_id:
            viewers[tick] = row
        else:
            enemies[tick].append(row)

    candidates: dict[int, list[dict[str, Any]]] = defaultdict(list)
    visibility_inputs: list[dict[str, Any]] = []
    visibility_targets: list[dict[str, Any]] = []
    numeric_id = int(steam_id)
    for tick, viewer in viewers.items():
        if float(viewer.get("health") or 0) <= 0:
            continue
        for enemy in enemies.get(tick, []):
            if float(enemy.get("health") or 0) <= 0 or enemy.get("side") == viewer.get(
                "side"
            ):
                continue
            candidate = {
                "tick": tick,
                "enemy_steamid": player_id(enemy.get("steamid")),
                "viewer_velocity": float(viewer.get("velocity") or 0.0),
                "target_x": enemy.get("X"),
                "target_y": enemy.get("Y"),
                "target_z": enemy.get("Z"),
                "target_duck": enemy.get("duck_amount"),
                "radar_visible": numeric_id
                in (enemy.get("approximate_spotted_by") or []),
            }
            visibility_inputs.append(
                {
                    **candidate,
                    "viewer_x": viewer.get("X"),
                    "viewer_y": viewer.get("Y"),
                    "viewer_z": viewer.get("Z"),
                    "viewer_duck": viewer.get("duck_amount"),
                    "viewer_pitch": viewer.get("pitch"),
                    "viewer_yaw": viewer.get("yaw"),
                }
            )
            visibility_targets.append(candidate)

    geometry_visibility = (
        visible_rows(visibility_inputs, mesh) if mesh is not None else []
    )
    for index, candidate in enumerate(visibility_targets):
        candidate["visible"] = (
            geometry_visibility[index]
            if mesh is not None
            else candidate["radar_visible"]
        )
        candidates[int(candidate["tick"])].append(candidate)
    return candidates


def closest_target(
    shot: dict[str, Any], candidates: list[dict[str, Any]]
) -> dict[str, Any] | None:
    origin = shot_origin(shot)
    direction = shot_direction(shot)
    if origin is None or direction is None:
        return None
    ranked: list[tuple[float, dict[str, Any]]] = []
    for candidate in candidates:
        if not candidate.get("visible"):
            continue
        target_direction = normalize(vector_subtract(target_head(candidate), origin))
        if target_direction is None:
            continue
        ranked.append((angular_distance(direction, target_direction), candidate))
    if not ranked:
        return None
    angle, target = min(ranked, key=lambda item: item[0])
    return target if angle <= MAX_TARGET_ANGLE else None


def calculate_spray_bursts(
    demo: Any, steam_id: str, mesh: Any = None
) -> list[dict[str, Any]]:
    ticks = getattr(demo, "ticks", None)
    if not isinstance(ticks, pl.DataFrame) or ticks.is_empty():
        return []
    grouped = group_sprays(fire_bullet_rows(demo, steam_id))
    shot_ticks = {
        int(shot.get("tick") or 0) for _weapon, burst in grouped for shot in burst
    }
    candidates = target_candidates(ticks, steam_id, shot_ticks, mesh)
    sprays: list[dict[str, Any]] = []

    for weapon, burst in grouped:
        points: list[dict[str, Any]] = []
        for number, shot in enumerate(burst, start=1):
            tick = int(shot.get("tick") or 0)
            target = closest_target(shot, candidates.get(tick, []))
            if target is None:
                continue
            velocity = float(target.get("viewer_velocity") or 0.0)
            if not points and (
                not math.isfinite(velocity)
                or velocity >= float(weapon["maxSpeed"]) * 0.34
            ):
                break
            offset = projected_offset(shot, target)
            if offset is None:
                continue
            points.append(
                {
                    "number": number,
                    "x": round(offset[0], 2),
                    "y": round(offset[1], 2),
                }
            )
        if len(points) >= MIN_SPRAY_SHOTS:
            sprays.append({"weapon": weapon["id"], "shots": points})
    return sprays


def percentile(values: list[float], fraction: float) -> float:
    ordered = sorted(values)
    if not ordered:
        return 0.0
    if len(ordered) == 1:
        return ordered[0]
    position = fraction * (len(ordered) - 1)
    lower = math.floor(position)
    upper = math.ceil(position)
    if lower == upper:
        return ordered[lower]
    weight = position - lower
    return ordered[lower] * (1.0 - weight) + ordered[upper] * weight


def confidence_label(spray_count: int) -> str:
    if spray_count >= 15:
        return "HIGH"
    if spray_count >= 8:
        return "GOOD"
    if spray_count >= 4:
        return "LOW"
    return "MORE DATA NEEDED"


def coach_spray(weapon_name: str, shots: list[dict[str, Any]], spray_count: int) -> str:
    if spray_count < 4:
        return f"Only {spray_count} qualifying {weapon_name} spray{'s' if spray_count != 1 else ''}; keep collecting matches."
    minimum_samples = max(2, math.ceil(spray_count / 2))
    late = [
        shot
        for shot in shots
        if 6 <= int(shot.get("number") or 0) <= 10
        and int(shot.get("samples") or 0) >= minimum_samples
    ]
    late = late or [
        shot for shot in shots if int(shot.get("samples") or 0) >= minimum_samples
    ]
    horizontal = statistics.median(float(shot["x"]) for shot in late)
    vertical = statistics.median(float(shot["y"]) for shot in late)
    prefix = "Early read: " if spray_count < 8 else ""
    suffix = " More sprays will firm this up." if spray_count < 8 else ""
    if abs(horizontal) < 7.0 and abs(vertical) < 7.0:
        return (
            prefix
            + "the later bullets stay centred; this spray shape is controlled."
            + suffix
        )
    if abs(horizontal) >= abs(vertical):
        if horizontal > 0:
            return (
                prefix
                + "the later bullets drift right; pull slightly further left after bullet 5."
                + suffix
            )
        return (
            prefix
            + "the later bullets drift left; ease the leftward pull after bullet 5."
            + suffix
        )
    if vertical > 0:
        return (
            prefix
            + "the later bullets climb high; pull down more firmly after bullet 5."
            + suffix
        )
    return (
        prefix
        + "the later bullets land low; ease the downward pull after bullet 5."
        + suffix
    )


def aggregate_sprays(matches: list[dict[str, Any]]) -> dict[str, Any]:
    bursts_by_weapon: dict[str, list[list[dict[str, Any]]]] = defaultdict(list)
    for match in matches[:10]:
        for burst in match.get("sprays", []):
            weapon = str(burst.get("weapon") or "")
            shots = burst.get("shots") or []
            if weapon in WEAPONS_BY_ID and len(shots) >= MIN_SPRAY_SHOTS:
                bursts_by_weapon[weapon].append(shots[:MAX_SPRAY_SHOTS])

    weapons: list[dict[str, Any]] = []
    for weapon_id in WEAPON_ORDER:
        weapon = WEAPONS_BY_ID[weapon_id]
        bursts = bursts_by_weapon.get(weapon_id, [])
        aggregated: list[dict[str, Any]] = []
        for number in range(1, MAX_SPRAY_SHOTS + 1):
            points = [
                shot
                for burst in bursts
                for shot in burst
                if int(shot.get("number") or 0) == number
            ]
            if not points:
                continue
            horizontal = [float(point["x"]) for point in points]
            vertical = [float(point["y"]) for point in points]
            aggregated.append(
                {
                    "number": number,
                    "x": round(statistics.median(horizontal), 2),
                    "y": round(statistics.median(vertical), 2),
                    "radiusX": round(
                        max(
                            1.5,
                            (
                                percentile(horizontal, 0.75)
                                - percentile(horizontal, 0.25)
                            )
                            / 2,
                        ),
                        2,
                    ),
                    "radiusY": round(
                        max(
                            1.5,
                            (percentile(vertical, 0.75) - percentile(vertical, 0.25))
                            / 2,
                        ),
                        2,
                    ),
                    "samples": len(points),
                }
            )
        spray_count = len(bursts)
        weapons.append(
            {
                "id": weapon_id,
                "name": weapon["name"],
                "shortName": weapon["shortName"],
                "sprays": spray_count,
                "confidence": confidence_label(spray_count),
                "shots": aggregated,
                "coach": coach_spray(weapon["name"], aggregated, spray_count),
            }
        )
    return {"matches": min(10, len(matches)), "weapons": weapons}

File: crates/omarcs/src/spray.rs
use std::collections::{HashMap, HashSet};

use serde::Serialize;

use crate::match_facts::{BulletFact, MatchFacts, PlayerId};

const SPRAY_GAP_TICKS: i32 = 16;
const MIN_SPRAY_SHOTS: usize = 5;
const MAX_SPRAY_SHOTS: usize = 10;
const MAX_TARGET_ANGLE: f64 = 15.0;
const MAX_PROJECTED_OFFSET: f64 = 220.0;
const SETTLED_SPEED_FRACTION: f32 = 0.34;

const WEAPONS: &[(i32, SprayWeapon)] = &[
    (
        7,
        SprayWeapon {
            id: "ak47",
            max_speed: 215.0,
        },
    ),
    (
        13,
        SprayWeapon {
            id: "galilar",
            max_speed: 215.0,
        },
    ),
    (
        16,
        SprayWeapon {
            id: "m4a4",
            max_speed: 225.0,
        },
    ),
    (
        60,
        SprayWeapon {
            id: "m4a1_silencer",
            max_speed: 225.0,
        },
    ),
];

#[derive(Clone, Copy)]
struct SprayWeapon {
    id: &'static str,
    max_speed: f32,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SprayShot {
    pub number: usize,
    pub x: f64,
    pub y: f64,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SprayBurst {
    pub weapon: &'static str,
    pub shots: Vec<SprayShot>,
}

struct TargetCandidate {
    viewer_velocity: f32,
    target_origin: [f32; 3],
    target_duck: f32,
    visible: bool,
}

pub fn calculate(facts: &MatchFacts, player: PlayerId) -> Vec<SprayBurst> {
    if facts.ticks.is_empty() {
        return Vec::new();
    }

    let grouped = group_sprays(
        facts
            .bullets
            .iter()
            .filter(|bullet| bullet.player == Some(player)),
    );
    let shot_ticks = grouped
        .iter()
        .flat_map(|(_, burst)| burst.iter().map(|shot| shot.tick))
        .collect::<HashSet<_>>();
    let candidates = target_candidates(facts, player, &shot_ticks);
    let mut sprays = Vec::new();

    for (weapon, burst) in grouped {
        let mut points = Vec::new();
        for (number, shot) in burst.into_iter().enumerate() {
            let number = number + 1;
            let Some(target) =
                closest_target(shot, candidates.get(&shot.tick).map_or(&[], Vec::as_slice))
            else {
                continue;
            };
            if points.is_empty()
                && (!target.viewer_velocity.is_finite()
                    || target.viewer_velocity >= weapon.max_speed * SETTLED_SPEED_FRACTION)
            {
                break;
            }
            let Some((x, y)) = projected_offset(shot, target) else {
                continue;
            };
            points.push(SprayShot {
                number,
                x: round_to(x, 2),
                y: round_to(y, 2),
            });
        }
        if points.len() >= MIN_SPRAY_SHOTS {
            sprays.push(SprayBurst {
                weapon: weapon.id,
                shots: points,
            });
        }
    }
    sprays
}

fn group_sprays<'a>(
    bullets: impl IntoIterator<Item = &'a BulletFact>,
) -> Vec<(SprayWeapon, Vec<&'a BulletFact>)> {
    let mut bullets = bullets.into_iter().collect::<Vec<_>>();
    bullets.sort_by_key(|bullet| bullet.tick);

    let mut grouped = Vec::new();
    let mut current: Vec<&BulletFact> = Vec::new();
    let mut current_weapon: Option<SprayWeapon> = None;
    let mut previous_tick: Option<i32> = None;

    for bullet in bullets {
        let Some(item_definition) = bullet.item_definition else {
            finish_group(&mut grouped, &mut current, &mut current_weapon);
            previous_tick = None;
            continue;
        };
        let Some(weapon) = weapon(item_definition) else {
            finish_group(&mut grouped, &mut current, &mut current_weapon);
            previous_tick = None;
            continue;
        };
        let continues = current_weapon.is_some_and(|current| current.id == weapon.id)
            && previous_tick.is_some_and(|previous| bullet.tick - previous <= SPRAY_GAP_TICKS);
        if !continues {
            finish_group(&mut grouped, &mut current, &mut current_weapon);
            current_weapon = Some(weapon);
        }
        current.push(bullet);
        previous_tick = Some(bullet.tick);
    }
    finish_group(&mut grouped, &mut current, &mut current_weapon);
    grouped
}

fn finish_group<'a>(
    grouped: &mut Vec<(SprayWeapon, Vec<&'a BulletFact>)>,
    current: &mut Vec<&'a BulletFact>,
    current_weapon: &mut Option<SprayWeapon>,
) {
    if let Some(weapon) = current_weapon.take()
        && current.len() >= MIN_SPRAY_SHOTS
    {
        current.truncate(MAX_SPRAY_SHOTS);
        grouped.push((weapon, std::mem::take(current)));
    } else {
        current.clear();
    }
}

fn target_candidates(
    facts: &MatchFacts,
    player: PlayerId,
    shot_ticks: &HashSet<i32>,
) -> HashMap<i32, Vec<TargetCandidate>> {
    let ticks = &facts.ticks;
    let mut viewers = HashMap::<i32, usize>::new();
    let mut enemies = HashMap::<i32, Vec<usize>>::new();
    for index in 0..ticks.len() {
        let tick = ticks.tick(index);
        if !shot_ticks.contains(&tick) {
            continue;
        }
        if ticks.player(index) == player {
            viewers.insert(tick, index);
        } else {
            enemies.entry(tick).or_default().push(index);
        }
    }

    let mut candidates = HashMap::<i32, Vec<TargetCandidate>>::new();
    for (tick, viewer_index) in viewers {
        if ticks.health(viewer_index) <= 0 {
            continue;
        }
        for &enemy_index in enemies.get(&tick).map(Vec::as_slice).unwrap_or(&[]) {
            if ticks.health(enemy_index) <= 0 || ticks.side(enemy_index) == ticks.side(viewer_index)
            {
                continue;
            }
            candidates.entry(tick).or_default().push(TargetCandidate {
                viewer_velocity: ticks.velocity(viewer_index),
                target_origin: ticks.origin(enemy_index),
                target_duck: ticks.duck_amount(enemy_index),
                visible: ticks.spotted_by(enemy_index).contains(&player),
            });
        }
    }
    candidates
}

fn closest_target<'a>(
    shot: &BulletFact,
    candidates: &'a [TargetCandidate],
) -> Option<&'a TargetCandidate> {
    let origin = shot_origin(shot)?;
    let direction = shot_direction(shot)?;
    let mut best: Option<(f64, &TargetCandidate)> = None;
    for candidate in candidates {
        if !candidate.visible {
            continue;
        }
        let Some(target_direction) = normalize(sub(target_head(candidate), origin)) else {
            continue;
        };
        let angle = angular_distance(direction, target_direction);
        if best.is_none_or(|(best_angle, _)| angle < best_angle) {
            best = Some((angle, candidate));
        }
    }
    best.and_then(|(angle, target)| (angle <= MAX_TARGET_ANGLE).then_some(target))
}

fn projected_offset(shot: &BulletFact, target: &TargetCandidate) -> Option<(f64, f64)> {
    let origin = shot_origin(shot)?;
    let direction = shot_direction(shot)?;
    let head = target_head(target);
    let target_vector = sub(head, origin);
    let normal = normalize(target_vector)?;
    let denominator = dot(direction, normal);
    if denominator <= 0.05 {
        return None;
    }
    let impact = add_scaled(origin, direction, length(target_vector) / denominator);
    let offset = sub(impact, head);
    let right = normalize([-normal[1], normal[0], 0.0])?;
    let up = normalize(cross(normal, right))?;
    let horizontal = dot(offset, right);
    let vertical = dot(offset, up);
    if !horizontal.is_finite() || !vertical.is_finite() {
        return None;
    }
    if horizontal.abs() > MAX_PROJECTED_OFFSET || vertical.abs() > MAX_PROJECTED_OFFSET {
        return None;
    }
    Some((horizontal, vertical))
}

fn shot_origin(shot: &BulletFact) -> Option<[f64; 3]> {
    let origin = [
        f64::from(shot.origin[0]),
        f64::from(shot.origin[1]),
        f64::from(shot.origin[2]),
    ];
    origin
        .iter()
        .all(|value| value.is_finite())
        .then_some(origin)
}

fn shot_direction(shot: &BulletFact) -> Option<[f64; 3]> {
    let pitch = f64::from(shot.angles[0]);
    let yaw = f64::from(shot.angles[1]);
    if !pitch.is_finite() || !yaw.is_finite() {
        return None;
    }
    Some(direction_from_angles(pitch, yaw))
}

fn target_head(target: &TargetCandidate) -> [f64; 3] {
    [
        f64::from(target.target_origin[0]),
        f64::from(target.target_origin[1]),
        f64::from(target.target_origin[2]) + 64.0 - 18.0 * f64::from(target.target_duck),
    ]
}

fn direction_from_angles(pitch_degrees: f64, yaw_degrees: f64) -> [f64; 3] {
    let pitch = pitch_degrees.to_radians();
    let yaw = yaw_degrees.to_radians();
    [
        pitch.cos() * yaw.cos(),
        pitch.cos() * yaw.sin(),
        -pitch.sin(),
    ]
}

fn angular_distance(first: [f64; 3], second: [f64; 3]) -> f64 {
    dot(first, second).clamp(-1.0, 1.0).acos().to_degrees()
}

fn weapon(item_definition: i32) -> Option<SprayWeapon> {
    WEAPONS
        .iter()
        .find_map(|(id, weapon)| (*id == item_definition).then_some(*weapon))
}

fn dot(first: [f64; 3], second: [f64; 3]) -> f64 {
    first[0] * second[0] + first[1] * second[1] + first[2] * second[2]
}

fn sub(first: [f64; 3], second: [f64; 3]) -> [f64; 3] {
    [
        first[0] - second[0],
        first[1] - second[1],
        first[2] - second[2],
    ]
}

fn add_scaled(origin: [f64; 3], direction: [f64; 3], scale: f64) -> [f64; 3] {
    [
        origin[0] + direction[0] * scale,
        origin[1] + direction[1] * scale,
        origin[2] + direction[2] * scale,
    ]
}

fn length(vector: [f64; 3]) -> f64 {
    dot(vector, vector).sqrt()
}

fn normalize(vector: [f64; 3]) -> Option<[f64; 3]> {
    let length = length(vector);
    (length > 1e-9).then(|| [vector[0] / length, vector[1] / length, vector[2] / length])
}

fn cross(first: [f64; 3], second: [f64; 3]) -> [f64; 3] {
    [
        first[1] * second[2] - first[2] * second[1],
        first[2] * second[0] - first[0] * second[2],
        first[0] * second[1] - first[1] * second[0],
    ]
}

fn round_to(value: f64, places: i32) -> f64 {
    let factor = 10_f64.powi(places);
    (value * factor).round() / factor
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::match_facts::{PlayerFact, RoundFact, Side};
    use crate::ticks::{TickObservations, TickSample};

    const PLAYER: PlayerId = PlayerId(76_561_198_000_000_001);
    const ENEMY: PlayerId = PlayerId(76_561_198_000_000_002);

    fn round() -> RoundFact {
        RoundFact {
            number: 1,
            start_tick: 1,
            freeze_end_tick: None,
            end_tick: 200,
            official_end_tick: 200,
            winner: Side::CounterTerrorist,
        }
    }

    fn sample(
        tick: i32,
        player: PlayerId,
        side: Side,
        origin: [f32; 3],
        yaw: f32,
        spotted_by: Vec<PlayerId>,
    ) -> TickSample {
        TickSample {
            tick,
            player,
            side,
            health: 100,
            origin,
            pitch: 0.0,
            yaw,
            duck_amount: 0.0,
            velocity: 0.0,
            shots_fired: 0,
            weapon: "ak47".to_owned(),
            spotted_by,
        }
    }

    fn bullet(tick: i32, yaw: f32) -> BulletFact {
        BulletFact {
            tick,
            player: Some(PLAYER),
            item_definition: Some(7),
            origin: [0.0, 0.0, 64.0],
            angles: [0.0, yaw, 0.0],
        }
    }

    fn fixture() -> MatchFacts {
        let rounds = vec![round()];
        let mut ticks = Vec::new();
        let mut bullets = Vec::new();
        for (number, tick) in [100, 106, 112, 118, 124, 130].into_iter().enumerate() {
            ticks.push(sample(
                tick,
                PLAYER,
                Side::CounterTerrorist,
                [0.0, 0.0, 0.0],
                0.0,
                vec![],
            ));
            ticks.push(sample(
                tick,
                ENEMY,
                Side::Terrorist,
                [100.0, 0.0, 0.0],
                180.0,
                vec![PLAYER],
            ));
            bullets.push(bullet(tick, number as f32));
        }
        MatchFacts {
            map: "de_test".to_owned(),
            players: vec![
                PlayerFact {
                    steam_id: PLAYER,
                    name: "Kieren".to_owned(),
                },
                PlayerFact {
                    steam_id: ENEMY,
                    name: "Enemy".to_owned(),
                },
            ],
            rounds,
            kills: vec![],
            damages: vec![],
            blinds: vec![],
            shots: vec![],
            bullets,
            ticks: TickObservations::from_samples(ticks, &[round()]),
            tick_rows: 0,
        }
    }

    #[test]
    fn projects_shot_onto_enemy_head_plane() {
        let target = TargetCandidate {
            viewer_velocity: 0.0,
            target_origin: [100.0, 0.0, 0.0],
            target_duck: 0.0,
            visible: true,
        };
        let centred = bullet(100, 0.0);
        let right = bullet(100, 5.0);

        assert_eq!(projected_offset(&centred, &target), Some((0.0, 0.0)));
        let (horizontal, vertical) = projected_offset(&right, &target).expect("offset");
        assert_eq!((horizontal * 10.0).round() / 10.0, 8.7);
        assert_eq!((vertical * 10.0).round() / 10.0, 0.0);
    }

    #[test]
    fn matches_python_target_relative_spray_fixture() {
        let sprays = calculate(&fixture(), PLAYER);
        assert_eq!(sprays.len(), 1);
        assert_eq!(sprays[0].weapon, "ak47");
        assert_eq!(
            sprays[0]
                .shots
                .iter()
                .map(|shot| shot.number)
                .collect::<Vec<_>>(),
            vec![1, 2, 3, 4, 5, 6]
        );
        assert_eq!(sprays[0].shots[0].x, 0.0);
        assert!(sprays[0].shots[5].x > 8.0);
    }

    #[test]
    fn returns_no_sprays_without_tick_observations() {
        let mut facts = fixture();
        facts.ticks = TickObservations::from_samples([], &facts.rounds);
        assert_eq!(calculate(&facts, PLAYER), Vec::<SprayBurst>::new());
    }
}
